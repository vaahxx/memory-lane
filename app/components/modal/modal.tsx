"use client";

import React, { useId, useRef } from "react";
import { Button, type ButtonProps } from "../button/button";

type ModalProps = {
  children: JSX.Element;
  triggerButtonProps?: ButtonProps;
  open: boolean;
  onModalStateChange: (open: boolean) => void;
};

export function Modal({
  children,
  triggerButtonProps,
  open,
  onModalStateChange,
}: ModalProps) {
  const id = useId();
  const childrenRef = useRef<HTMLDivElement>(null);

  const handleOutsideClick = (
    event: React.MouseEvent<HTMLDivElement, MouseEvent>
  ) => {
    const isChildElement =
      event.target instanceof Node &&
      childrenRef.current?.contains(event.target as Node);

    if (!isChildElement) {
      onModalStateChange(false);
    }
  };

  return (
    <>
      <Button
        type='button'
        onClick={() => {
          onModalStateChange(!open);
        }}
        {...triggerButtonProps}
      />

      {open && (
        <div
          id={id}
          onClick={handleOutsideClick}
          className='flex overflow-y-auto overflow-x-hidden fixed top-0 right-0 left-0 z-50 justify-center items-center w-full md:inset-0 h-[calc(100%-1rem)] max-h-full bg-black bg-opacity-50'
        >
          <div className='bg-white p-10 rounded-md w-3/4' ref={childrenRef}>
            {children}
          </div>
        </div>
      )}
    </>
  );
}
