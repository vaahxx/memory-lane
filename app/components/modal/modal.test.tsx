import React from "react";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import { Modal } from "./modal";

describe("Modal Component", () => {
  it("renders the Modal component with trigger button", () => {
    render(
      <Modal open onModalStateChange={() => {}}>
        <div data-testid='modal-content'>Modal Content</div>
      </Modal>
    );

    expect(screen.getByTestId("modal-content")).toMatchSnapshot();
  });

  it("opens the modal when the trigger button is clicked", async () => {
    const onModalStateChangeMock = jest.fn();
    render(
      <Modal open onModalStateChange={onModalStateChangeMock}>
        <div>Modal Content</div>
      </Modal>
    );

    fireEvent.click(screen.getByRole("button"));

    await waitFor(() => {
      expect(onModalStateChangeMock).toHaveBeenCalled();
    });
  });

  it("closes the modal when clicking outside the modal content", async () => {
    const onModalStateChangeMock = jest.fn();
    const { container } = render(
      <Modal open onModalStateChange={onModalStateChangeMock}>
        <div>Modal Content</div>
      </Modal>
    );

    fireEvent.click(container.firstChild!);

    await waitFor(() => {
      expect(onModalStateChangeMock).toHaveBeenCalledWith(false);
    });
  });
});
