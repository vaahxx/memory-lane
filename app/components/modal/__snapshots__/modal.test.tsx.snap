// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component renders the Modal component with trigger button 1`] = `
<div
  data-testid="modal-content"
>
  Modal Content
</div>
`;
